### PR TITLE
Refactor Interop validation into separate phase

### PR DESCRIPTION
**Pull request by warunalakshitha**
## Purpose
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/26771

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/CompilerDriver.java [hunk 1]
```diff
@@ -389,6 +403,6 @@ private BPackageSymbol getLangModuleFromSource(PackageID modID) {
             return null;
         }
 
-        return codeGen(birGen(desugar(pkg))).symbol;
+        return codeGen(birEmit(birGen(desugar(pkg)))).symbol;
```
Comment (warunalakshitha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/CompilerDriver.java:406]: Do we need interopvalidate here. Codegen function is empty with new changes.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmPackageGen.java [hunk 2]
```diff
@@ -596,9 +591,7 @@ private void linkGlobalVars(BIRPackage module, String pkgName, String initClass,
         globalVarClassMap.put(pkgName + LOCK_STORE_VAR_NAME, initClass);
     }
 
-    private void linkTypeDefinitions(BIRPackage module, String pkgName,
-                                     InteropValidator interopValidator,
-                                     boolean isEntry, String orgName,
+    private void linkTypeDefinitions(BIRPackage module, String pkgName, boolean isEntry, String orgName,
```
Comment (Kishanthan) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmPackageGen.java:594]: Can we check whether pkgName and orgName would be same values that can be used from `BIRPackage module` here?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmPackageGen.java [hunk 3]
```diff
@@ -545,13 +543,10 @@ private boolean listenerDeclarationFound(List<BIRGlobalVariableDcl> variableDcls
...
      * @param module           bir module
      * @param initClass        module init class name
-     * @param interopValidator interop validator instance
      * @param isEntry          is entry module flag
      * @return The map of javaClass records on given source file name
      */
-    private Map<String, JavaClass> generateClassNameLinking(BIRPackage module, String initClass,
-                                                            InteropValidator interopValidator,
-                                                            boolean isEntry) {
+    private Map<String, JavaClass> generateClassNameLinking(BIRPackage module, String initClass, boolean isEntry) {
```
Comment (Kishanthan) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmPackageGen.java:549]: Any way we could check and see whether we can get rid of this flag `isEntry` and have separate functions.? That given better readability IMO.

Comment (warunalakshitha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmPackageGen.java:548]: isEntry is passed from generate module to each function generation. This is only used for add default Boolean params to function signature. So code will be duplicate and much complex code if try to separate them to functions. IMO correct thing is to move all desugar code (including encodes, record inits) from codegen to another class which needs to be done separately.

**Review by manuranga — APPROVED** [review]

**Pull request merged**